modules/ROOT/pages/architecture/hexagonale_architecture.adoc

```diff
@@ -0,0 +1,140 @@
+//        - mapper
+//          - ReservationMapper.java
+//     - out
+//       - jpa
+//         - .java
+//         - model
+//           - ReservationEntity.java
+//           - TableEntity.java
+//         - mapper
+//           - ReservationJpaMapper.java
+//           - TableJpaMapper.java
+// ----
+
+[source,plaintext]
+----
+application/
+├── config/
+│   └── Configuration.java
+├── core/
+│   ├── domain/
+│   │   ├── Customer.java
+│   │   ├── CustomerFactory.java
+│   │   ├── Reservation.java
+│   │   └── Table.java
+│   ├── ports/
+│   │   └── StoreReservationPort.java
+│   ├── service/
+│   │   └── FindTableService.java
+│   └── usecase/
+│       ├── AddReservationUc.java
+│       └── AddTableUc.java
+└── adapter/
+    ├── in/
+    │   └── rest
+    │       ├── RestController.java
+    │       ├── model/
+    │       │   └── ReservationDto.java
+    │       └── mapper/
+    │           └── ReservationMapper.java
+    └── out/
+        └── jpa/
+            ├── .java
+            ├── model/
+            │   ├── ReservationEntity.java
+            │   └── TableEntity.java
+            └── mapper/
+                ├── ReservationJpaMapper.java
+                └── TableJpaMapper.java
+----
+
+[cols="1,1", options="header"]
+|===
+| Package | Description
+
+| core
+| The core contains the essential business logic, domain entities, and use cases. It focuses on implementing the main functionalities while remaining technology-agnostic. The core interacts with external components through well-defined interfaces called "ports," ensuring a clear separation of concerns and promoting flexibility, testability, and maintainability.
+
+| core.domain
+| The domain circle contains the entities and value objects of the business domain of the application. 
+Related Factories or Builders are located here as well. 
+The entities in our domain are usually rich. 
+
+| core.service
+| Services provide a reusable part of the applications business logic that is used by multiple use cases.
+
+| core.usecase
+| Use Cases are the main entrypoint of the applications core. They validate the given input and orchestrate the domain entities, services and ports to implement a Business Use Case.
+
+| core.port
+| Ports are interfaces, that are used by the core and should be implemented by an according adapter. Ports should not be technology specific. One big advantage of the hexagonal architecture is, that the adapters can be changed without changing the core and therefore, without touching the business logic.
+
+| adapter
+a| Adapters connect the application core to the surrounding context. They have the following tasks:
+
+* Implement a specific protocol to connect to the context. E.g REST, JDBC, MQTT, ...
+* Maintain a data model that is necessary to communicate with the context
+* Translate the domain model from the core to that model or vice versa
+* Handle protocol specific errors
+* Log the interaction with the surrounding context
+
+| adapter.in
```

Comment:
Actually there is a fine but significant difference between in and out adapters. The "in" adapters are the ones driving your application, in case of Kafka being an "in" adapter, that would most likely be the API of your application, via which external requests are admitted into the system, while the "out" adapter would be the implementations of the ports declared in your domain, to enable your domain to interact with external system.

For example the "in" adapter would be a class annotated with `@KafkaListener` annotation, while the "out" adapter for Kafka would be the implementation of the interface which enables the domain to write events to the Kafka topics. 

It is very unlikely for an adapter to be both "in" and "out", since the difference is that "in" adapters ( or otherwise called the driving adapters) are the ones initiating interactions in your domain core, while the "out" adapters ( or otherwise called the driven adapters ) are invoked by your application.

Reply:
I totally agree for this differentiation on class or package level. However if you separate your application into different modules (e.g. maven) it's most probably not necessary to have two different modules for communication with Kafka. Of course inside the module it should be separated into different packages / classes.

Reply:
Moving this discussion to #42